### DISCUSSION TITLE
PIP-23: Sign and verify message by public and private key account

### DISCUSSION THREAD

**Discussion by ja7ad** [pactus-project/PIPs#102] [PIPs]
PIP-23: [Sign and verify message by public and private key account](https://pips.pactus.org/PIPs/pip-23)

**Comment by ja7ad** [comment]
@b00f @kehiy @ambersun1234 Please review this PIP for accepting to implementation.

**Reply by b00f** [reply]
@Ja7ad 
This PIP should be categorized under "Interface". It adds new APIs to the gRPC. 
It should cover 3 methods:

- `SignMessageWithPrivateKey` (Util Service)
- `VerifyMessage`  (Util Service)
- `SignMessage`  (Wallet Service)

For the `SignMessage`, user needs to provide address and wallet password.

**Reply by kehiy** [reply]
@Ja7ad @b00f

I think these small changes were for Ethereum and Bitcoin when there was no path and they were in an evolution.

Currently, we have well-defined protocols to inherit from them or use them. which supports this logic and a lot more.

We can use Chain Agnostic improvement proposals Wallet Connect standard:
https://walletconnect.com/

Or implement a wallet connect standard like TON Connect specified for Pactus:
https://github.com/ton-blockchain/TEPs/blob/master/text/0115-ton-connect.md


Both of them has pros and cons!
But adding 3 functions and improving it step by step is reinventing the wheel.

If you are agree, I think we can disscuess in this way to have one of these. and change pip-23 status to withdrawn.

**Reply by kehiy** [reply]
@Ja7ad @b00f

I think these small changes were for Ethereum and Bitcoin when there was no path and they were in an evolution.

Currently, we have well-defined protocols to inherit from them or use them. which supports this logic and a lot more.

We can use Chain Agnostic improvement proposals Wallet Connect standard:
https://walletconnect.com/

Or implement a wallet connect standard like TON Connect specified for Pactus:
https://github.com/ton-blockchain/TEPs/blob/master/text/0115-ton-connect.md


Both of them have pros and cons!
But adding 3 functions and improving it step by step is reinventing the wheel.

If you agree, I think we can discuss in this way to have one of these. and change pip-23 status to withdrawn.

**Reply by b00f** [reply]
No, Please let's don't mix things.
This PIP is not categorized as"Wallet". It is related to "Interface" .

For that purpose we work on new PIP

**Reply by kehiy** [reply]
@b00f why we should have PIP for a simple code improvement? which is done a lot of times before?

**Reply by b00f** [reply]
@Ja7ad please update the PIP to and explain these APIs.



> @Ja7ad This PIP should be categorized under "Interface". It adds new APIs to the gRPC. It should cover 3 methods:
> 
>     * `SignMessageWithPrivateKey` (Util Service)
> 
>     * `VerifyMessage`  (Util Service)
> 
>     * `SignMessage`  (Wallet Service)
> 
> 
> For the `SignMessage`, user needs to provide address and wallet password.

**Reply by ja7ad** [reply]
> @Ja7ad @b00f
> 
> I think these small changes were for Ethereum and Bitcoin when there was no path and they were in an evolution.
> 
> Currently, we have well-defined protocols to inherit from them or use them. which supports this logic and a lot more.
> 
> We can use Chain Agnostic improvement proposals Wallet Connect standard: https://walletconnect.com/
> 
> Or implement a wallet connect standard like TON Connect specified for Pactus: https://github.com/ton-blockchain/TEPs/blob/master/text/0115-ton-connect.md
> 
> Both of them have pros and cons! But adding 3 functions and improving it step by step is reinventing the wheel.
> 
> If you agree, I think we can discuss in this way to have one of these. and change pip-23 status to withdrawn.

walletconnect used sign and verify message for connecting then give some json-rpc method to interaction Dapps with wallet.

This is low layer methodology to sign and verify custom message with private and public key.

**Comment by b00f** [comment]
There are some issues with the PIP yet:

1. Private key is not in hex, it is in bech32 encoding.
    "Input: Message as string, Private key associated with Pactus account."
2. For "Example" we should provide data. Not function signature that is explained above. We can remove Example section.
3. Services are not mentioned. We can update like:

* `SignMessageWithPrivateKey` (Util Service)
* `VerifyMessage`  (Util Service)
* `SignMessage`  (Wallet Service)

**Reply by ja7ad** [reply]
> Private key is not in hex, it is in bech32 encoding.

What is it result of bench32 encode?

> For "Example" we should provide data. Not function signature that is explained above. We can remove Example section.

Ok

> Services are not mentioned. We can update like.

I add service name for this methods.

**Reply by b00f** [reply]
Yes, Private Keys encoded as string by bech32m, Example: https://github.com/pactus-project/pactus/blob/08559b417be1f22eab80c0cf812a35af3e640854/crypto/bls/private_key_test.go#L88

**Reply by ja7ad** [reply]
> Yes, Private Keys encoded as string by bech32m, Example: https://github.com/pactus-project/pactus/blob/08559b417be1f22eab80c0cf812a35af3e640854/crypto/bls/private_key_test.go#L88

Thanks understood

**Reply by ja7ad** [reply]
@b00f Please review: https://github.com/pactus-project/PIPs/pull/149

**Comment by ja7ad** [comment]
Example diagram: https://t.ly/-wSwu

**Comment by ja7ad** [comment]
Issue implementation created: https://github.com/pactus-project/pactus/issues/1386

**Comment by ja7ad** [comment]
This PIP implemented in: https://github.com/pactus-project/pactus/pull/1397
